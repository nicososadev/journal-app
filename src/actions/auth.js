import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types'

export const loginManager = (email, password) => {

    return (dispatch) => {

        setTimeout(() => {
            
            dispatch( login(123, 'Juan') )
        }, 3500);
    }
}

export const googleLoginManager = () => {
    
    return (dispatch) => {

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then( ({ user }) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            })
    }
}

export const login = (uid, displayName) => {

    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}