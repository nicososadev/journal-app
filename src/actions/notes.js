import Swal from "sweetalert2"
import { db } from "../firebase/firebase-config"
import { fileUpload } from "../helpers/fileUpload"
import { loadNotes } from "../helpers/loadNotes"
import { types } from "../types/types"

export const createNote = () => {

    return async(dispatch, getState) => {

        const state = getState()

        const { uid } = state.auth

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const document = await db.collection(`${uid}/journal/notes`).add(newNote)

        dispatch( activeNote(document.id, newNote) )

        dispatch( createNewNote(document.id, newNote) )

    }
}

export const activeNote = (id, note) => {

    return {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
}

export const createNewNote = (id, note) => {

    return {
        type: types.notesCreateNote,
        payload: {
            id,
            ...note
        }
    }
}

export const LoadNotesManager = (uid) => {

    return async (dispatch) => {
        
        const notes = await loadNotes( uid )

        dispatch( setNotes(notes) )
    }
}

export const setNotes = (notes) => {

    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const saveNoteManager = ( note ) => {

    return async (dispatch, getState) => {
        
        const state = getState()

        const { uid } = state.auth

        if( !note.url ){
            delete note.url
        }

        const noteToFireStore = { ...note }

        delete noteToFireStore.id

        await db.doc(`${uid}/journal/notes/${note.id}`).update( noteToFireStore )

        dispatch( noteRefresh( note.id, noteToFireStore ) )

        Swal.fire('Saved', note.title, 'success')
    }
}

export const noteRefresh = ( id, note ) => {

    return {
        type: types.notesUpdateNote,
        payload: {
            id, 
            note: {
                id,
                ...note
            }
        }
    }
}

export const uploadManager = ( file ) => {

    return async (dispatch, getState) => {

        const { active:activeNote } = getState().notes

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading()
            }
        })
        

        const fileUrl = await fileUpload( file )

        activeNote.url = fileUrl

        dispatch( saveNoteManager( activeNote ) )

        Swal.close()
    }
}

export const deleteManager = ( id ) => {

    return async (dispatch, getState) => {

        const state = getState()

        const { uid } = state.auth

        await db.doc(`${uid}/journal/notes/${id}`).delete()

        dispatch( deleteNote( id ) )
    }
}

export const deleteNote = ( id ) => {
    return {
        type: types.notesDeleteNote,
        payload: id
    }
}

export const notesLogout = () => {
    return{
        type: types.notesLogoutCleaning
    }
}