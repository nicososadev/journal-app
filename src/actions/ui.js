import { types } from "../types/types"

export const setError = ( error ) => {
   
    return {
        type: types.uiSetError,
        payload: error
    }
}
