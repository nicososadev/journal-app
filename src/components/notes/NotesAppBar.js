import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { saveNoteManager } from '../../actions/notes'

export const NotesAppBar = () => {

    const dispatch = useDispatch()

    const { active } = useSelector(state => state.notes)

    const handleSaveNote = () => {

        dispatch( saveNoteManager( active ) )
    }

    return (

        <div className="notes__appbar">
            <span>28 de agosto 2020</span>
            <div>
                <button className="btn">Picture</button>
                <button onClick={handleSaveNote} className="btn">Save</button>
            </div>
        </div>
    )
}
