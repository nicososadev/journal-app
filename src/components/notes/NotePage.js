import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, deleteManager } from '../../actions/notes'
import { useForm } from '../../hooks/useForm'
import { NotesAppBar } from './NotesAppBar'

export const NotePage = () => {

    const dispatch = useDispatch()

    const { active:note } = useSelector(state => state.notes)

    const [ formValues, handleInputChange, reset ] = useForm( note )

    const { title, body } = formValues

    const activeId = useRef( note.id )

    useEffect(() => {

        if( note.id !== activeId.current ){
            reset( note )
            activeId.current = note.id
        }

    }, [note, reset])

    useEffect(() => {
        
        dispatch( activeNote( formValues.id, { ...formValues } ) )

    }, [formValues, dispatch])

    const handleDelete = () => {

        dispatch( deleteManager( activeId.current ) )
    }

    return (

        <div className="notes__main-content">

            <NotesAppBar />

            <div className="notes__content">
                <input type="text" className="notes__title-input" value={title} name="title" onChange={handleInputChange} placeholder="Some awesome title" />
                <hr className="mb-5" />
                <textarea className="notes__textarea" value={body} name="body" onChange={handleInputChange} placeholder="An awesome text">

                </textarea>
                
                {
                    (note.url) &&
                    <div className="notes__image">
                        <img src={note.url} alt="imagen"/>
                    </div>
                }
            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>
        </div>

    )
}
