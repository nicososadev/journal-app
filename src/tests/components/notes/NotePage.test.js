import '@testing-library/jest-dom'
import React from 'react'
import thunk from 'redux-thunk'
import configureStore from 'redux-mock-store'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import { NotePage } from '../../../components/notes/NotePage'
import { activeNote } from '../../../actions/notes'

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn()
}))

// jest.mock('../../../actions/notes', () => ({
//    createNote: jest.fn()
// }))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialState = {
    auth: {
        uid: '12345',
        displayName: 'Nicolas'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: {
            id: 1234,
            title: 'Prueba',
            body: 'Nota de Prueba'
        }
    }
}

let store = mockStore(initialState)

store.dispatch = jest.fn()

describe('Pruebas en el componente NotePage', () => {
    
    const wrapper = mount(
        <Provider store={store}>
            <NotePage/>
        </Provider>
    )

    test('Debe mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot()
    })
    
    test('Debe lanzar la accion activeNote', () => {
        
        wrapper.find('input[name="title"]').simulate('change', {
            target: {
                name: 'title',
                value: 'Modificando titulo'
            }
        })

        expect( activeNote ).toHaveBeenLastCalledWith(
            1234,
            {
                body: 'Nota de Prueba',
                title: 'Modificando titulo',
                id: 1234,
            }
        )
    }) 
})
