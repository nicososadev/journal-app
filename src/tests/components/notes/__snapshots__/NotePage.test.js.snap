// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente NotePage Debe mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 de agosto 2020
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="Prueba"
    />
    <hr
      className="mb-5"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder="An awesome text"
      value="Nota de Prueba"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
